src/main/java/net/neoforged/neoforge/attachment/AttachmentType.java

```diff
@@ -58,165 +53,131 @@
-            }
-        });
+    public static <T, P extends IAttachmentHolder> Builder<T, P> builder(Class<P> parentClass, Function<P, T> defaultValueConstructor) {
+        return new Builder<>(defaultValueConstructor);
     }
 
-    public static class Builder<T> {
-        private final Function<IAttachmentHolder, T> defaultValueSupplier;
+    public static class Builder<T, P extends IAttachmentHolder> {
+        private final Function<P, T> defaultValueSupplier;
+        @Nullable
+        private Encoder<T> serializer;
         @Nullable
-        private IAttachmentSerializer<?, T> serializer;
+        private Decoder<T> deserializer;
+        BiConsumer<P, T> postDeserialize;
         private boolean copyOnDeath;
+        private Predicate<T> shouldSerialize;
         @Nullable
-        private IAttachmentCopyHandler<T> copyHandler;
+        private StreamCodec<RegistryFriendlyByteBuf, T> copyHandler;
 
-        private Builder(Function<IAttachmentHolder, T> defaultValueSupplier) {
+        private Builder(Function<P, T> defaultValueSupplier) {
             this.defaultValueSupplier = defaultValueSupplier;
+            this.shouldSerialize = Predicates.alwaysTrue();
+            this.postDeserialize = (holder, inst) -> {};
         }
 
         /**
-         * Requests that this attachment be persisted to disk (on the logical server side).
+         * Requests that this attachment be persisted to disk (on the logical server side), using a {@link Codec}.
          *
-         * @param serializer The serializer to use.
+         * @param codec The codec to use.
          */
-        public Builder<T> serialize(IAttachmentSerializer<?, T> serializer) {
-            Objects.requireNonNull(serializer);
+        public Builder<T, P> serialize(Encoder<T> codec) {
+            Objects.requireNonNull(codec);
             if (this.serializer != null)
                 throw new IllegalStateException("Serializer already set");
 
-            this.serializer = serializer;
+            this.serializer = codec;
             return this;
         }
 
         /**
-         * Requests that this attachment be persisted to disk (on the logical server side), using a {@link Codec}.
+         * For attachments that are serialized to disk, specifies a different codec to use for deserialization.
          *
-         * <p>Using a {@link Codec} to serialize attachments is discouraged for item stack attachments,
-         * for performance reasons. Prefer one of the other options.
-         *
-         * <p>Codec-based attachments cannot capture a reference to their holder.
-         *
-         * @param codec The codec to use.
+         * @param deserializeCodec A codec to use to deserialize attachment data.
          */
-        public Builder<T> serialize(Codec<T> codec) {
-            return serialize(codec, Predicates.alwaysTrue());
+        public Builder<T, P> deserialize(Decoder<T> deserializeCodec) {
+            if (this.serializer == null)
```

Comment:
Why does this have to be called _after_ `serialize`? That seems rather weird. Just have it throw on build if the deserializer is set but not the serializer.